# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
step_four_data = pd.read_csv('step3_output.csv')
step_four_data

Name       state  summit_elev  vertical_drop  \
0              Eaglecrest Ski Area      Alaska         2600           1540   
1                 Hilltop Ski Area      Alaska         2090            294   
2              Sunrise Park Resort     Arizona        11100           1800   
3    Yosemite Ski & Snowboard Area  California         7800            600   
4           Boreal Mountain Resort  California         7700            500   
..                             ...         ...          ...            ...   
152                    Tyrol Basin   Wisconsin         1160            300   
153              Whitecap Mountain   Wisconsin         1750            400   
154                  Hogadon Basin     Wyoming         8000            640   
155      Sleeping Giant Ski Resort     Wyoming         7428            810   
156               Snow King Resort     Wyoming         7808           1571   

     trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0        0        0.0          0          0     0       0  ...   
1        0        0.0          0          0     0       1  ...   
2        0        0.0          0          1     2       3  ...   
3        0        0.0          0          0     0       1  ...   
4        0        0.0          0          1     1       3  ...   
..     ...        ...        ...        ...   ...     ...  ...   
152      0        0.0          0          0     0       3  ...   
153      0        0.0          0          0     1       0  ...   
154      0        0.0          0          0     0       0  ...   
155      0        0.0          0          0     0       1  ...   
156      0        0.0          0          0     1       1  ...   

     SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0                640.0            60.0              45.0       44.0   
1                 30.0            30.0             150.0       36.0   
2                800.0            80.0             115.0       49.0   
3                 88.0             0.0             110.0       84.0   
4                380.0           200.0             150.0       54.0   
..                 ...             ...               ...        ...   
152               32.0            32.0             112.0       61.0   
153              400.0           300.0             105.0       57.0   
154               92.0            32.0             121.0       61.0   
155              184.0            18.0              61.0       81.0   
156              400.0           250.0             121.0       80.0   

     averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0              350.0          47.0      53.00000               90.0   
1               69.0          30.0      34.00000              152.0   
2              250.0          74.0      78.00000              104.0   
3              300.0          47.0      47.00000              107.0   
4              400.0          49.0      64.16681              150.0   
..               ...           ...           ...                ...   
152             41.0          17.0      48.00000              103.0   
153            200.0          60.0      60.00000              118.0   
154             80.0          48.0      48.00000               95.0   
155            310.0          42.0      42.00000               77.0   
156            300.0          59.0      59.00000              123.0   

     NightSkiing_ac  clusters  
0               0.0         0  
1              30.0         1  
2              80.0         2  
3               0.0         2  
4             200.0         2  
..              ...       ...  
152            32.0         1  
153             0.0         1  
154             0.0         2  
155             0.0         2  
156           110.0         2  

[157 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
state_dummies = pd.get_dummies(step_four_data['state'])
step_four_data = pd.concat([step_four_data.drop(['state'], axis=1), state_dummies], axis=1)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [4]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = step_four_data.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = step_four_data.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [5]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [6]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [7]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [8]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn import metrics
mean_absolute_error = metrics.mean_absolute_error(y_test,y_pred)
explained_variance_score = metrics.explained_variance_score(y_test,y_pred)
print('Mean Absolute Error',mean_absolute_error)
print('Explained Variance Score',explained_variance_score)

Mean Absolute Error 43526079826.15707
Explained Variance Score -9.117160939204048e+19


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [9]:
model.intercept_

-11089447030.484318

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [12]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coeff_dataframe = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coeff_dataframe.sort_values(by='Coefficient',ascending=True).head(30)

Coefficient
South Dakota       2.197266e-03
Snow Making_ac     1.069336e-01
LongestRun_mi      2.287598e-01
TerrainParks       5.041504e-01
daysOpenLastYear   5.979004e-01
projectedDaysOpen  6.364746e-01
Runs               8.666992e-01
averageSnowfall    8.912354e-01
clusters           9.545898e-01
NightSkiing_ac     1.040894e+00
vertical_drop      1.114383e+00
yearsOpen          1.149902e+00
SkiableTerrain_ac  1.182007e+00
Rhode Island       2.681268e+00
summit_elev        4.977886e+00
AdultWeekday       9.400269e+00
Maryland           2.471505e+06
Arizona            4.616823e+10
Iowa               4.616823e+10
Missouri           4.616823e+10
Tennessee          4.616823e+10
Montana            6.508213e+10
Nevada             6.508213e+10
Indiana            6.508213e+10
Alaska             6.508213e+10
Virginia           6.508213e+10
Wyoming            7.945146e+10
Oregon             7.945146e+10
Utah               7.945146e+10
Washington         7.945146e+10

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [13]:
model_2_data = pd.read_csv('step3_output.csv')
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = model_2_data.drop(['Name','state','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = model_2_data.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

In [14]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
mean_absolute_error = metrics.mean_absolute_error(y_test,y_pred)
explained_variance_score = metrics.explained_variance_score(y_test,y_pred)
print('Mean Absolute Error',mean_absolute_error)
print('Explained Variance Score',explained_variance_score)

Mean Absolute Error 43526079826.15707
Explained Variance Score -9.117160939204048e+19


In [17]:
coeff_dataframe = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coeff_dataframe.sort_values(by='Coefficient',ascending=True).head(100)

Coefficient
fastSixes          4.440892e-16
trams              8.881784e-16
fastEight          8.881784e-16
projectedDaysOpen  3.768803e-02
surface            4.657670e-02
Runs               6.579511e-02
double             1.698004e-01
LongestRun_mi      1.845421e-01
Snow Making_ac     2.840763e-01
daysOpenLastYear   2.929246e-01
fastQuads          2.976304e-01
NightSkiing_ac     6.435661e-01
yearsOpen          6.510468e-01
total_chairs       6.888547e-01
averageSnowfall    8.122467e-01
summit_elev        8.492082e-01
TerrainParks       1.006257e+00
triple             1.145081e+00
quad               1.186498e+00
SkiableTerrain_ac  1.211129e+00
vertical_drop      2.431038e+00
clusters           3.288346e+00
AdultWeekday       1.059971e+01

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [18]:
model_3_data = pd.read_csv('step3_output.csv')
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = model_3_data.drop(['Name','state','summit_elev','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = model_3_data.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

In [21]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
mean_absolute_error = metrics.mean_absolute_error(y_test,y_pred)
explained_variance_score = metrics.explained_variance_score(y_test,y_pred)
print('Mean Absolute Error',mean_absolute_error)
print('Explained Variance Score',explained_variance_score)

Mean Absolute Error 43526079826.15707
Explained Variance Score -9.117160939204048e+19


In [23]:
coeff_dataframe = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coeff_dataframe.sort_values(by='Coefficient',ascending=True).head(23)

Coefficient
fastSixes          4.440892e-16
fastEight          6.661338e-16
trams              8.881784e-16
Runs               9.453111e-02
surface            9.475494e-02
projectedDaysOpen  9.970406e-02
Snow Making_ac     1.609914e-01
double             1.854084e-01
daysOpenLastYear   2.134904e-01
fastQuads          2.503085e-01
LongestRun_mi      2.551595e-01
NightSkiing_ac     5.397298e-01
yearsOpen          6.555798e-01
total_chairs       6.579047e-01
averageSnowfall    6.730915e-01
TerrainParks       1.037375e+00
triple             1.171213e+00
SkiableTerrain_ac  1.242747e+00
quad               1.296972e+00
vertical_drop      2.679455e+00
clusters           2.839085e+00
AdultWeekday       1.066379e+01

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -9.117160939204048e+19 |43526079826.15707  |-|
| Model 2. | -9.117160939204048e+19 |43526079826.15707 |'state'|
| Model 3. | -9.117160939204048e+19 |43526079826.15707 |'state','summit_elev','base_elev'|

Model Selection: